File: python-selenium-pytest-bdd/src/saucedemo_uitests/testdata/checkout_data_factory.py
"""Generates realistic checkout form data with Faker. Login credentials are never generated this way -
see `users.user_catalog` - only data the app accepts freely.

Builds a fresh `Faker` instance on every call rather than sharing one at module scope: unlike the C# and
TS siblings' thread-based parallelism, pytest-xdist workers are separate processes (see
`users.assigned_user`), so there is no in-process concurrent access to guard against here - a fresh
instance is simply the easiest way to make the optional `seed` parameter instance-local (via
`seed_instance`) without ever touching `Faker.seed`, which reseeds the shared generator *every* Faker
instance draws from (confirmed directly: it is a classmethod, the exact global-mutable-state trap the C#
sibling's own comment warns about for `Bogus.Randomizer.Seed`).
"""

from faker import Faker

from saucedemo_uitests.testdata.checkout_info import CheckoutInfo


def create_checkout_info(seed: int | None = None) -> CheckoutInfo:
    fake = Faker()
    if seed is not None:
        fake.seed_instance(seed)

    return CheckoutInfo(
        first_name=fake.first_name(),
        last_name=fake.last_name(),
        postal_code=fake.zipcode(),
    )

File: ts-playwright-test/src/testdata/checkoutDataFactory.ts
import { faker } from '@faker-js/faker';
import type { CheckoutInfo } from './checkoutInfo';

/**
 * Generates realistic checkout form data with Faker. Login credentials are never generated this way -
 * see `userCatalog.ts` - only data the app accepts freely.
 *
 * Uses faker-js's shared singleton, which is safe here for the same reason it is in the Cucumber
 * sibling, and if anything more clearly so: a Playwright worker is a separate OS *process* (not a
 * worker thread), with its own module registry and its own copy of this module, running one test at a
 * time. There is no shared-memory concurrency for a per-call `Faker<T>` instance to protect against,
 * which is why the two C# siblings need one and this does not.
 */
export function createCheckoutInfo(): CheckoutInfo {
  return {
    firstName: faker.person.firstName(),
    lastName: faker.person.lastName(),
    postalCode: faker.location.zipCode('#####')
  };
}
